2005-11-02  Andrew Pinski  <pinskia@physics.uc.edu>

  PR fortran/18157
  * trans-array.c (gfc_conv_resolve_dependencies): Use the
  correct type for the temporary array.
  * trans-expr.c (gfc_trans_assignment): Pass
  lss instead of lss_section to gfc_conv_resolve_dependencies
  to get the correct type.

--- gcc/fortran/trans-array.c	2005/11/02 21:01:54	106396
+++ gcc/fortran/trans-array.c	2005/11/02 21:12:27	106397
@@ -2379,10 +2379,13 @@
 
   if (nDepend == 1)
     {
+      tree base_type = gfc_typenode_for_spec (&dest->expr->ts);
+      if (GFC_ARRAY_TYPE_P (base_type)
+	  || GFC_DESCRIPTOR_TYPE_P (base_type))
+	base_type = gfc_get_element_type (base_type);
       loop->temp_ss = gfc_get_ss ();
       loop->temp_ss->type = GFC_SS_TEMP;
-      loop->temp_ss->data.temp.type =
-	gfc_get_element_type (TREE_TYPE (dest->data.info.descriptor));
+      loop->temp_ss->data.temp.type = base_type;
       loop->temp_ss->string_length = dest->string_length;
       loop->temp_ss->data.temp.dimen = loop->dimen;
       loop->temp_ss->next = gfc_ss_terminator;
--- gcc/fortran/trans-expr.c	2005/11/02 21:01:54	106396
+++ gcc/fortran/trans-expr.c	2005/11/02 21:12:27	106397
@@ -2293,7 +2293,7 @@
       /* Calculate the bounds of the scalarization.  */
       gfc_conv_ss_startstride (&loop);
       /* Resolve any data dependencies in the statement.  */
-      gfc_conv_resolve_dependencies (&loop, lss_section, rss);
+      gfc_conv_resolve_dependencies (&loop, lss, rss);
       /* Setup the scalarizing loops.  */
       gfc_conv_loop_setup (&loop);
 
